File: algorithms/remove_duplicates_from_sorted_list.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def delete_duplicates(head)
  while head
    while head.next and head.next.val == head.val
      head.next = head.next.next
    end
    ans ||= head
    head = head.next
  end
  ans
end

a = ListNode.new(1)
b = ListNode.new(1)

a.next = b

p delete_duplicates(a)

File: algorithms/remove_duplicates_from_sorted_list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        p = head
        while p is not None:
            while p.next is not None and \
                  p.val == p.next.val:
                p.next = p.next.next
            p = p.next
        return head

t = Solution()

a = ListNode(1)
b = ListNode(1)
c = ListNode(1)
a.next = b
b.next = c

t.deleteDuplicates(a)

assert a.next == None

print("tests passed")
